Gitlab: Normalize comments and issues access

shulard opened the pull request:
The Gitlab API transform all the retrieved data into objects and :

* It was impossible to view issue detail
* It was impossible to view issue comments

This PR normalize the same way than on the others thirdparty libs.

pierredup commented on src/ThirdParty/Gitlab/Adapter/GitLabIssueTracker.php:
```diff
@@ -29,7 +29,7 @@ class GitLabIssueTracker extends BaseIssueTracker
     public function openIssue($subject, $body, array $options = [])
     {
-        if (isset($options['assignee'])) {
+        if (isset($options['assignee']) && !empty($options['assignee'])) {
```
I think just doing the !empty should be enough here

shulard replied:
You are right, `empty` is enough here.

pierredup commented on src/ThirdParty/Gitlab/Adapter/GitLabIssueTracker.php:
```diff
@@ -80,7 +86,7 @@ public function getIssueUrl($id)
             $this->getUsername(),
             $this->getRepository(),
-            $this->getIssue($id)['iid']
+            ($id instanceof Issue)?$id->iid:$this->getIssue($id)['iid']
```
Looks like a typo here (iid)

sstok replied:
Missing spaces between operators, we use the PSR-2 coding style.

shulard replied:
Ok, I've ran php-cs-fixer here but I think it missed this particular case...
Updated !

pierredup commented on src/ThirdParty/Gitlab/Adapter/GitLabIssueTracker.php:
```diff
@@ -190,7 +196,18 @@ public function getComments($id)
 
-        return $issue->showComments();
+        $comments = [];
+        $issue->showComments();
```
This looks a bit weird standing alone without using the return value. Is this intended?

shulard replied:
I know that this seems a bit weird but I do not use `showComments` return value directly.

There is a problem in the `php-gitlab-api` lib which does not allow to access the `id` of a comment in the properties.
So I need to do the following steps :

* Execute the request to retrieve comment list with `showComments`
* Access to the `HttpClient` last response with `getLastResponse` call
* Extract and retrieve all the requested details

I think I'll submit a PR on the `php-gitlab-api` project but before the changes will be accepted, we need to deal with that ugly stuff.

pierredup replied:
Can't you access the data form the `getData` method on the model?
E.G

```
$comments = array_map(function(Note $comment) {
    $data = $comment->getData();

    return [
        'id' => $data['id'],
        'user' => ['login' => $data['author']['username']],
        'body' => $data['body'],
        'created_at' => new \DateTime($data['created_at']),
    ];
}, $issue->showComments());
```

(Note: I haven't tested this, but looking at the gitlab api code, you should be able to get all the values from the `getData` method)

pierredup replied:
Never mind, I see when setting the data on the model, it only sets the data that is available in the properties

shulard replied:
Hum, after reading the code of Gitlab API model more precisely it seems that you are right. Get data will simplify the code :) So the php-gitlab-api PR is not useless but not really necessary for our case.

shulard replied:
Ok you are right, the `setData` method is used in the `hydrate` one so we are blocked by the property check...

We now need to wait for the php-gitlab-api update to be published to simplify our code...

cordoval commented:
thanks @pierredup for reviewing, and the work @shulard

shulard commented:
I've created the PR on the `php-gitlab-api` project: m4tthumphrey/php-gitlab-api/pull/110

It'll allow to simplify the `getComments` code because the only missing thing is the access to `Note::id` property :smile:.

sstok commented on src/ThirdParty/Gitlab/Adapter/GitLabIssueTracker.php:
```diff
@@ -49,7 +49,7 @@ public function openIssue($subject, $body, array $options = [])
             [
                 'description' => $body,
-                'assignee_id' => isset($assigneeId) ? $assigneeId : null,
+                'assignee_id' => isset($assigneeId) ? $assigneeId : "",
```
Please use the `''` instead `""` whenever possible :+1:

m4tthumphrey commented:
Yo @shulard ! I've merged the PR :smile:

shulard commented:
Hello @m4tthumphrey, when do you think a release of php-gitlab-api can be done ? We need a new stable of this lib to merge this PR :smile:

cordoval commented:
@shulard lock it on composer to a sha1 and don't wait.

shulard commented:
Hello !

I've fixed the version of `php-gitlab-api` and rebased my PR to have a more clean history. Also the strange thing about `$issue->showComments()` is now fixed :smile:.

If everything is ok for you, I'll merge that PR !
